onnxruntime/contrib_ops/webgpu/quantization/matmul_nbits.cc

```diff
@@ -408,38 +270,62 @@ Status MatMulNBits::ComputeInternal(onnxruntime::webgpu::ComputeContext& context
 
   // WideTileProgram
   // This program is optimized for Block32 prefill using Tile16x128.
-  const bool use_wide_tile_program = block_size == 32 && components_a == 4 && components_b == 4 && M >= kMinMForTileOptimization;
+  const bool use_wide_tile_program = block_size == 32 &&
+                                     components_a == 4 &&
+                                     components_b == 4 &&
+                                     M >= kMinMForTileOptimization;
   if (use_wide_tile_program) {
     // Enforce output components to 1.
     components = 1;
 
     constexpr uint32_t workgroup_size = 128;
     constexpr uint32_t tile_m = workgroup_size / 8;
     constexpr uint32_t tile_n = workgroup_size;
-    uint32_t num_N_tile = (N + tile_n - 1) / tile_n;
-    uint32_t num_M_tile = (M + tile_m - 1) / tile_m;
+    const uint32_t num_N_tile = CEIL_DIV(N, tile_n);
+    const uint32_t num_M_tile = CEIL_DIV(M, tile_m);
 
     MatMulNBitsWideTileProgram program{has_zero_points, tile_m, tile_n, nbits};
     program.SetWorkgroupSize(workgroup_size);
-    program.SetDispatchGroupSize((N + tile_n - 1) / tile_n,
-                                 (M + tile_m - 1) / tile_m,
-                                 batch_count);
-    program.CacheHint("Tile" + std::to_string(tile_m) + "x" + std::to_string(tile_n) + "_Block32");
+    program.SetDispatchGroupSize(num_N_tile, num_M_tile, batch_count);
 
     TensorShape reshaped_a_shape{batch_count, M, K / components_a};
     TensorShape reshaped_b_shape{N, n_blocks_per_col, blob_size_in_words / components_b};
     TensorShape reshaped_y_shape{batch_count, M, N / components};
 
-    program
-        .AddInputs({{a, ProgramTensorMetadataDependency::TypeAndRank, reshaped_a_shape, onnxruntime::narrow<int>(components_a)},
-                    {b, ProgramTensorMetadataDependency::TypeAndRank, reshaped_b_shape, onnxruntime::narrow<int>(components_b * 4)},
-                    {scales, ProgramTensorMetadataDependency::None}})
-        .AddOutput({y, ProgramTensorMetadataDependency::TypeAndRank, reshaped_y_shape, onnxruntime::narrow<int>(components)})
-        .AddUniformVariables({{block_size}, {zero_blocks_per_col}, {num_N_tile}, {num_M_tile}})
-        .CacheHint(nbits, has_zero_points);
+    constexpr uint32_t kU32Components = 4;
+    const uint32_t components_b_with_u32 = components_b * kU32Components;
+    const uint32_t K_of_b = n_blocks_per_col * blob_size / components_b_with_u32;
+
+    program.AddInput({a,
+                      ProgramTensorMetadataDependency::TypeAndRank,
+                      reshaped_a_shape,
+                      onnxruntime::narrow<int>(components_a)});
```

Comment:
With your refactor, you even don't need a reshaped_xxx_shape, you can directly add the inputs like below:
```
  program
      .AddInputs({{a, ProgramTensorMetadataDependency::TypeAndRank, static_cast<int>(components_a)},
                  {b, ProgramTensorMetadataDependency::TypeAndRank, static_cast<int>(components_b_with_u32)},
                  {scales, ProgramTensorMetadataDependency::None}});